src/app/lightning/get-lightning-fee.ts

```diff
@@ -0,0 +1,119 @@
+import { CachedRouteLookupKeyFactory } from "@domain/routes/key-factory"
+import { LndService } from "@services/lnd"
+import { RoutesCache } from "@services/redis/routes"
+
+export const lnInvoiceProbeForFee = async ({
+  paymentRequest,
+}: {
+  paymentRequest: EncodedPaymentRequest
+}): Promise<Satoshis | ApplicationError> => {
+  const decodedInvoice = await decodeInvoice(paymentRequest)
+  if (decodedInvoice instanceof Error) return decodedInvoice
+  const paymentAmount = checkedToSats(decodedInvoice.amount || 0)
+  if (paymentAmount instanceof Error) return paymentAmount
+
+  return feeProbe({ decodedInvoice, paymentAmount })
+}
+
+export const lnNoAmountInvoiceProbeForFee = async ({
+  paymentRequest,
+  amount,
+}: {
+  paymentRequest: EncodedPaymentRequest
+  amount: Satoshis
+}): Promise<Satoshis | ApplicationError> => {
+  const decodedInvoice = await decodeInvoice(paymentRequest)
+  if (decodedInvoice instanceof Error) return decodedInvoice
+
+  const { amount: lnInvoiceAmount } = decodedInvoice
+  if (lnInvoiceAmount && lnInvoiceAmount > 0) {
+    return new LnPaymentRequestZeroAmountRequiredError()
+  }
+
+  const paymentAmount = checkedToSats(amount)
+  if (paymentAmount instanceof Error) return paymentAmount
+
+  return noAmountProbeForFee({ decodedInvoice, paymentAmount })
+}
+
+const feeProbe = async ({
+  decodedInvoice,
+  paymentAmount,
+}: {
+  decodedInvoice: LnInvoice
+  paymentAmount: Satoshis
+}): Promise<Satoshis | ApplicationError> => {
+  const { destination, paymentHash, milliSatsAmount } = decodedInvoice
+
+  const lndService = LndService()
+  if (lndService instanceof Error) throw lndService
+  if (lndService.isLocal(destination)) {
+    return toSats(0)
+  }
+
+  const key = CachedRouteLookupKeyFactory().create({
+    paymentHash,
+    milliSats: milliSatsAmount,
+  })
+  const routeFromCache = await RoutesCache().findByKey(key)
+  const validCachedRoute = !(routeFromCache instanceof Error)
+  if (validCachedRoute) return toSats(routeFromCache.route.fee)
+
+  const maxFee = LnFeeCalculator().max(paymentAmount)
+
+  const rawRoute = await lndService.invoiceProbeForRoute({ decodedInvoice, maxFee })
+  if (rawRoute instanceof Error) return rawRoute
+
+  const routeToCache = { pubkey: lndService.defaultPubkey(), route: rawRoute }
+  const cachedRoute = await RoutesCache().store({ key, routeToCache })
+  if (cachedRoute instanceof Error) return cachedRoute
+
+  return toSats(rawRoute.fee)
+}
+
+const noAmountProbeForFee = async ({
+  decodedInvoice,
+  paymentAmount,
+}: {
+  decodedInvoice: LnInvoice
+  paymentAmount: Satoshis
+}): Promise<Satoshis | ApplicationError> => {
+  const { destination, paymentHash } = decodedInvoice
+
+  const lndService = LndService()
+  if (lndService instanceof Error) throw lndService
+  if (lndService.isLocal(destination)) {
+    return toSats(0)
+  }
+
+  const key = CachedRouteLookupKeyFactory().create({
+    paymentHash,
+    milliSats: toMilliSatsFromNumber(paymentAmount),
```

Comment:
is this the only difference from `lnNoAmountInvoiceProbeForFee`?

Reply:
This and the call to `lndService.noAmountInvoiceProbeForRoute`. I started de-duping but these two differences are probably enough of a difference to signal separate use-cases.

The heuristic we've been using is to not try to de-dupe separate use-cases, and to not de-dupe until the 3rd duplication of code (guided by @bodymindarts)

Reply:
De-dupe discussion resolved in dev call yesterday